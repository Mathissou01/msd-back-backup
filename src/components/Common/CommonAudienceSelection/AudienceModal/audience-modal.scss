@use "src/styles/01-settings/settings.colors" as colors;

.c-AudienceModal {
  display: flex;
  flex-direction: column;
  gap: 32px;
  align-items: stretch;

  &__ModalTitle {
    color: colors.$expert-blue;
    font-weight: 700;
    font-size: 28px;
    line-height: 35px;
  }

  &__ModalFields {
    padding-top: 16px;
    color: colors.$gray-600;
    font-weight: 400;
    font-size: 14px;
    line-height: 21px;
  }

  &__ModalButtons {
    display: flex;
    flex-direction: row;
    gap: 24px;
  }
}
