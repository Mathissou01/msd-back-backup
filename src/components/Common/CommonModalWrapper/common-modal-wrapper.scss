@use "src/styles/01-settings/settings.colors" as colors;
@use "src/styles/02-tools/tools.shadows" as shadows;

.c-CommonModalWrapper {
  height: 100%;
  overflow: hidden;

  &:not(:empty) {
    position: absolute;
    top: 0;
    z-index: 10;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100vw;
    height: 100vh;
    padding: 20px 0;
    background-color: rgb(0 0 0 / 50%);
  }

  &__Modal {
    @include shadows.shadow(2);

    position: relative;
    display: flex;
    flex-direction: column;
    padding: 36px;
    background-color: colors.$white;
    border-radius: 10px;

    &_sizeMedium {
      width: 58%;
    }

    &_sizeLarge {
      width: 95%;
    }
  }

  &__Close {
    position: absolute;
    top: 15px;
    right: 15px;
    width: 40px;
    height: 40px;
    background: url("data:image/svg+xml,%3Csvg width='24' height='24' viewBox='0 0 24 24' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='m.75 23.249 22.5-22.5m0 22.5L.75.749' stroke='%23030F40' stroke-width='1.5' stroke-linecap='round' stroke-linejoin='round'/%3E%3C/svg%3E");
    background-repeat: no-repeat;
    background-position: center;
    background-size: 23px;
  }

  hgroup {
    margin-bottom: 30px;
  }
}
