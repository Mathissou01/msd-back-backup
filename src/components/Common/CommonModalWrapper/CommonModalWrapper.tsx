import classNames from "classnames";
import React, {
  forwardRef,
  ReactNode,
  useCallback,
  useEffect,
  useId,
  useImperativeHandle,
  useRef,
  useState,
} from "react";
import { createPortal } from "react-dom";
import { FocusFirstElement } from "../../../lib/utilities";
import "./common-modal-wrapper.scss";

export enum ICommonModalWrapperSize {
  MEDIUM = "medium",
  LARGE = "large",
}

interface ICommonModalWrapperProps {
  onClose?: () => void;
  children: ReactNode;
  size?: ICommonModalWrapperSize;
}

export type CommonModalWrapperRef = {
  toggleModal: (setValue?: boolean, isCancel?: boolean) => void;
};

export default forwardRef(function CommonModalWrapper(
  {
    onClose,
    children,
    size = ICommonModalWrapperSize.MEDIUM,
  }: ICommonModalWrapperProps,
  parentRef,
) {
  const modalId = useId();
  const ref = useRef<Element | null>(null);
  const [isMounted, setMounted] = useState<boolean>(false);

  const handleToggle = useCallback(
    (setValue?: boolean) => {
      ref.current = document.querySelector<HTMLElement>("#modal-portal");
      const newValue = setValue ?? !isMounted;
      setMounted(newValue);
      if (!newValue && onClose) onClose();
    },
    [isMounted, onClose],
  );

  useEffect(() => {
    function handleEscapeKey(event: KeyboardEvent) {
      if (event.code === "Escape") {
        handleToggle(false);
      }
    }

    document.addEventListener("keydown", handleEscapeKey);
    return () => document.removeEventListener("keydown", handleEscapeKey);
  }, [handleToggle]);

  useImperativeHandle(
    parentRef,
    () => ({
      toggleModal(setValue?: boolean) {
        handleToggle(setValue);
      },
    }),
    [handleToggle],
  );

  const focusRef = useCallback((node: HTMLDivElement) => {
    FocusFirstElement(node);
  }, []);

  function handleKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
    if (e.key !== "Tab" || !ref.current) return;
    const parentElement = ref.current?.querySelector(`[id="${modalId}"]`);
    const focusableModalElements = parentElement?.querySelectorAll(
      "a[href], button:not([disabled]), textarea, input, select",
    );
    const firstElement = focusableModalElements?.[0] as HTMLElement;
    const lastElement = focusableModalElements?.[
      focusableModalElements.length - 1
    ] as HTMLElement;

    if (!e.shiftKey && document.activeElement === lastElement) {
      firstElement.focus();
      return e.preventDefault();
    }

    if (e.shiftKey && document.activeElement === firstElement) {
      lastElement.focus();
      e.preventDefault();
    }
  }

  return isMounted && ref.current
    ? createPortal(
        <div
          className="c-CommonModalWrapper"
          onClick={() => handleToggle(false)}
          id={modalId}
          data-testid="common-modal-wrapper"
        >
          <div
            className={classNames("c-CommonModalWrapper__Modal", {
              "c-CommonModalWrapper__Modal_sizeMedium":
                size === ICommonModalWrapperSize.MEDIUM,
              "c-CommonModalWrapper__Modal_sizeLarge":
                size === ICommonModalWrapperSize.LARGE,
            })}
            onClick={(e) => e.stopPropagation()}
            onKeyDown={(e) => handleKeyDown(e)}
            ref={focusRef}
            data-testid="common-modal"
          >
            {children}
            <button
              className="c-CommonModalWrapper__Close"
              onClick={() => handleToggle(false)}
              data-testid="common-modal-close"
            />
          </div>
        </div>,
        ref.current,
      )
    : null;
});
