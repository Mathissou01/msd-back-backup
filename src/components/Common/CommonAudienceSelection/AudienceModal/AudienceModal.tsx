import { useEffect, useState } from "react";
import { useFormContext } from "react-hook-form";
import { useContract } from "../../../../hooks/useContract";
import { removeNulls } from "../../../../lib/utilities";
import CommonModalWrapper, {
  CommonModalWrapperRef,
} from "../../CommonModalWrapper/CommonModalWrapper";
import CommonButton from "../../CommonButton/CommonButton";
import FormSingleMultiselect, {
  IFormSingleMultiselectOption,
} from "../../../Form/FormSingleMultiselect/FormSingleMultiselect";
import "./audience-modal.scss";

interface IAudienceModalProps {
  modalRef: React.RefObject<CommonModalWrapperRef>;
  selectedAudiences?: Array<IFormSingleMultiselectOption>;
  onValidate: (audiences: Array<IFormSingleMultiselectOption>) => void;
}

export default function AudienceModal({
  modalRef,
  selectedAudiences,
  onValidate,
}: IAudienceModalProps) {
  /* Static Data */
  const labels = {
    title: "Sélectionner les usagers",
    submitBtn: "Valider la sélection",
    cancelBtn: "Annuler",
  };
  const formLabels = {
    users: "Usagers",
  };

  /* Methods */
  const handleCloseModal = () => {
    resetField("audiences");
    setValue("audiences", selectedAudiences);
  };

  function handleValidation() {
    onValidate(audiencesWatch);
    modalRef.current?.toggleModal(false);
    setValue("audiences", audiencesWatch, { shouldDirty: true });
  }

  /* Local Data */
  const { contract } = useContract();
  const [audienceOptions, setAudienceOptions] = useState<
    Array<IFormSingleMultiselectOption>
  >([]);
  const [isDisabled, setIsDisabled] = useState<boolean>(true);
  const { watch, setValue, resetField } = useFormContext();
  const audiencesWatch = watch("audiences", selectedAudiences);

  useEffect(() => {
    if (audienceOptions.length === 0) {
      const audiencesFromContract = contract?.attributes?.audiences?.data;
      if (audiencesFromContract) {
        setAudienceOptions(
          audiencesFromContract
            ?.map((audience) => {
              if (audience?.attributes?.type && audience.id) {
                return {
                  label: audience.attributes.type,
                  value: audience.id,
                };
              }
            })
            .filter(removeNulls),
        );
      }
    }
  }, [audienceOptions, contract]);

  useEffect(() => {
    if (audiencesWatch) {
      setIsDisabled(
        !Array.isArray(audiencesWatch) || audiencesWatch.length === 0,
      );
    }
  }, [audiencesWatch]);

  useEffect(() => {
    if (selectedAudiences) setValue("audiences", selectedAudiences);
  }, [selectedAudiences, setValue]);

  return (
    <CommonModalWrapper ref={modalRef} onClose={handleCloseModal}>
      <div className="c-AudienceModal__ModalTitle">{labels.title}</div>
      <hgroup>
        <div className="c-AudienceModal__ModalFields">
          <FormSingleMultiselect
            name="audiences"
            label={formLabels.users}
            options={audienceOptions}
            isMulti
            isRequired
          />
        </div>
      </hgroup>
      <div className="c-AudienceModal__ModalButtons">
        <CommonButton
          type="button"
          label={labels.submitBtn}
          style="primary"
          isDisabled={isDisabled}
          onClick={() => handleValidation()}
        />
        <CommonButton
          type="button"
          label={labels.cancelBtn}
          onClick={() => {
            modalRef.current?.toggleModal(false);
            handleCloseModal();
          }}
        />
      </div>
    </CommonModalWrapper>
  );
}
