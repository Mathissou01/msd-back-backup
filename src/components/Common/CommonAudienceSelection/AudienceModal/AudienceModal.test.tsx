import { act, fireEvent, render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import { CommonModalWrapperRef } from "../../CommonModalWrapper/CommonModalWrapper";
import AudienceModal from "./AudienceModal";

const twoAudiencesOptions = [
  { label: "Particuliers", value: 1 },
  { label: "Collectifs", value: 2 },
];
const individualsText = "Particuliers";
const collectivesText = "Collectifs";
const submitBtnLabel = "Valider la s√©lection";
const cancelBtnLabel = "Annuler";

const childRef = React.createRef<CommonModalWrapperRef>();

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <FormProvider {...formMethods}>
      {props.children}
      <div id="modal-portal" />
    </FormProvider>
  );
};

describe("AudienceModal", () => {
  it("renders", () => {
    const { container } = render(
      <Wrapper>
        <AudienceModal
          audienceOptions={twoAudiencesOptions}
          modalRef={childRef}
          onValidate={jest.fn()}
        />
      </Wrapper>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders with two selected audiences", async () => {
    const { container } = render(
      <Wrapper>
        <AudienceModal
          audienceOptions={twoAudiencesOptions}
          modalRef={childRef}
          onValidate={jest.fn()}
          selectedAudiences={twoAudiencesOptions}
        />
      </Wrapper>,
    );
    await act(() => {
      childRef.current?.toggleModal(true);
    });
    expect(await screen.findByText(individualsText)).toBeInTheDocument();
    expect(await screen.findByText(collectivesText)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("submits when clicking on submit button", async () => {
    const validateFunction = jest.fn();
    render(
      <Wrapper>
        <AudienceModal
          audienceOptions={twoAudiencesOptions}
          modalRef={childRef}
          onValidate={validateFunction}
          selectedAudiences={twoAudiencesOptions}
        />
      </Wrapper>,
    );
    await act(() => {
      childRef.current?.toggleModal(true);
    });
    expect(await screen.findByText(individualsText)).toBeInTheDocument();
    expect(await screen.findByText(collectivesText)).toBeInTheDocument();
    const submitButton = await screen.findByText(submitBtnLabel);
    expect(submitButton).toBeInTheDocument();
    await fireEvent.click(submitButton);
    expect(validateFunction).toHaveBeenCalled();
  });

  it("cancels when clicking on cancel button", async () => {
    render(
      <Wrapper>
        <AudienceModal
          audienceOptions={twoAudiencesOptions}
          modalRef={childRef}
          onValidate={jest.fn()}
          selectedAudiences={twoAudiencesOptions}
        />
      </Wrapper>,
    );
    await act(() => {
      childRef.current?.toggleModal(true);
    });
    expect(await screen.findByText(individualsText)).toBeInTheDocument();
    expect(await screen.findByText(collectivesText)).toBeInTheDocument();
    const cancelButton = await screen.findByText(cancelBtnLabel);
    expect(cancelButton).toBeInTheDocument();
    await fireEvent.click(cancelButton);
    expect(screen.queryByText(individualsText)).not.toBeInTheDocument();
    expect(screen.queryByText(collectivesText)).not.toBeInTheDocument();
  });
});
