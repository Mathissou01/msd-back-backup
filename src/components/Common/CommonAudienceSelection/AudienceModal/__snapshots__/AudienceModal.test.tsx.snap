// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudienceModal renders 1`] = `
<div>
  <div
    id="modal-portal"
  />
</div>
`;

exports[`AudienceModal renders with two selected audiences 1`] = `
<div>
  <div
    id="modal-portal"
  >
    <div
      class="c-CommonModalWrapper"
      data-testid="common-modal-wrapper"
      id=":r1:"
    >
      <div
        class="c-CommonModalWrapper__Modal c-CommonModalWrapper__Modal_sizeMedium"
        data-testid="common-modal"
      >
        <div
          class="c-AudienceModal"
        >
          <div
            class="c-AudienceModal__ModalTitle"
          >
            SÃ©lectionner les usagers
          </div>
          <hgroup>
            <div
              class="c-AudienceModal__ModalFields"
            >
              <div
                class="c-FormLabel"
              >
                <label
                  class="c-FormLabel__LabelWrapper"
                  for="audiences"
                >
                  <span
                    class="c-FormLabel__Label"
                  >
                    Usagers *
                  </span>
                </label>
                <div
                  class="c-FormLabel__Content"
                >
                  <div
                    class="c-FormSingleMultiselect__Input css-b62m3t-container"
                    id="audiences"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-2-live-region"
                    >
                      <span
                        id="aria-selection"
                      >
                        options Particuliers,Collectifs, selected.
                      </span>
                      <span
                        id="aria-context"
                      >
                          Select is focused ,type to refine list, press Down to open the menu,  press left to focus selected values
                      </span>
                    </span>
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="form-single-multiselect__control form-single-multiselect__control--is-focused css-t3ipsp-control"
                    >
                      <div
                        class="form-single-multiselect__value-container form-single-multiselect__value-container--is-multi form-single-multiselect__value-container--has-value css-3w2yfm-ValueContainer"
                      >
                        <div
                          class="form-single-multiselect__multi-value css-1p3m7a8-multiValue"
                        >
                          <div
                            class="form-single-multiselect__multi-value__label css-wsp0cs-MultiValueGeneric"
                          >
                            Particuliers
                          </div>
                          <div
                            aria-label="Remove Particuliers"
                            class="form-single-multiselect__multi-value__remove css-12a83d4-MultiValueRemove"
                            role="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="css-tj5bde-Svg"
                              focusable="false"
                              height="14"
                              viewBox="0 0 20 20"
                              width="14"
                            >
                              <path
                                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                              />
                            </svg>
                          </div>
                        </div>
                        <div
                          class="form-single-multiselect__multi-value css-1p3m7a8-multiValue"
                        >
                          <div
                            class="form-single-multiselect__multi-value__label css-wsp0cs-MultiValueGeneric"
                          >
                            Collectifs
                          </div>
                          <div
                            aria-label="Remove Collectifs"
                            class="form-single-multiselect__multi-value__remove css-12a83d4-MultiValueRemove"
                            role="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="css-tj5bde-Svg"
                              focusable="false"
                              height="14"
                              viewBox="0 0 20 20"
                              width="14"
                            >
                              <path
                                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                              />
                            </svg>
                          </div>
                        </div>
                        <div
                          class="form-single-multiselect__input-container css-qbdosj-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-describedby="react-select-2-live-region"
                            aria-errormessage="audiences_error"
                            aria-expanded="false"
                            aria-haspopup="true"
                            aria-invalid="false"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class="form-single-multiselect__input"
                            id="react-select-2-input"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class="form-single-multiselect__indicators css-1hb7zxy-IndicatorsContainer"
                      >
                        <div
                          aria-hidden="true"
                          class="form-single-multiselect__indicator form-single-multiselect__clear-indicator css-15lsz6c-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                            />
                          </svg>
                        </div>
                        <span
                          class="form-single-multiselect__indicator-separator css-1u9des2-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class="form-single-multiselect__indicator form-single-multiselect__dropdown-indicator css-15lsz6c-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <div>
                      <input
                        name="tags"
                        type="hidden"
                        value="1"
                      />
                      <input
                        name="tags"
                        type="hidden"
                        value="2"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </hgroup>
          <div
            class="c-AudienceModal__ModalButtons"
          >
            <button
              class="c-CommonButton c-CommonButton_primary c-CommonButton_fontSmall"
              type="button"
            >
              Valider la sÃ©lection
            </button>
            <button
              class="c-CommonButton c-CommonButton_secondary c-CommonButton_fontSmall"
              type="button"
            >
              Annuler
            </button>
          </div>
        </div>
        <button
          class="c-CommonModalWrapper__Close"
          data-testid="common-modal-close"
        />
      </div>
    </div>
  </div>
</div>
`;
